src/h3/mod.rs

```diff
@@ -26,4 +26,940 @@
+                        .unwrap(),
+                    http::header::HeaderValue::from_str(hdr.value()).unwrap(),
+                );
+            },
+        }
+    }
+
+    *resp.status_mut() = StatusCode::from_bytes(status.as_bytes()).unwrap();
+    *resp.version_mut() = http::Version::HTTP_2;
+
+    // debug!("Prepared response {:?}", resp);
+
+    resp
+}
+
+/// An HTTP/3 configuration.
+pub struct Config {
+    pub root_dir: String,
+    pub num_placeholders: u64,
+    pub max_header_list_size: u64,
+    pub qpack_max_table_capacity: u64,
+    pub qpack_blocked_streams: u64,
+}
+
+impl Config {
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new() -> Result<Config> {
+        Ok(Config {
+            root_dir: String::new(),
+            num_placeholders: 16,
+            max_header_list_size: 0,
+            qpack_max_table_capacity: 0,
+            qpack_blocked_streams: 0,
+        })
+    }
+
+    pub fn set_root_dir(&mut self, root_dir: &String) {
+        self.root_dir = String::clone(root_dir);
+    }
+
+    pub fn set_num_placeholders(&mut self, num_placeholders: u64) {
+        self.num_placeholders = num_placeholders;
+    }
+
+    pub fn set_max_header_list_size(&mut self, max_header_list_size: u64) {
+        self.max_header_list_size = max_header_list_size;
+    }
+
+    pub fn set_qpack_max_table_capacity(
+        &mut self, qpack_max_table_capacity: u64,
+    ) {
+        self.qpack_max_table_capacity = qpack_max_table_capacity;
+    }
+
+    pub fn set_qpacked_blocked_streams(&mut self, qpack_blocked_streams: u64) {
+        self.qpack_blocked_streams = qpack_blocked_streams;
+    }
+}
+
+type RequestMap = BTreeMap<u64, http::Request<()>>;
+type StreamMap = BTreeMap<u64, stream::Stream>;
+
+/// An HTTP/3 connection.
+pub struct Connection {
+    root_dir: String,
+    is_server: bool,
+
+    requests: RequestMap,
+    streams: StreamMap,
+    uni_stream_ledger: Vec<u64>,
+
+    num_placeholders: u64,
+    max_header_list_size: u64,
+    qpack_max_table_capacity: u64,
+    qpack_blocked_streams: u64,
+
+    peer_num_placeholders: Option<u64>,
+    peer_max_header_list_size: Option<u64>,
+    peer_qpack_max_table_capacity: Option<u64>,
+    peer_qpack_blocked_streams: Option<u64>,
+
+    control_stream_id: Option<u64>,
+    peer_control_stream_open: bool,
+
+    qpack_encoder: qpack::Encoder,
+    qpack_decoder: qpack::Decoder,
+
+    qpack_encoder_stream_open: bool,
+    peer_qpack_encoder_stream_open: bool,
+    qpack_decoder_stream_open: bool,
+    peer_qpack_decoder_stream_open: bool,
+}
+
+impl Connection {
+    #[allow(clippy::new_ret_no_self)]
+    fn new(config: &mut Config, is_server: bool) -> Result<Connection> {
+        let root = String::clone(&config.root_dir); // TODO shouldn't need to clone here
+
+        Ok(Connection {
+            root_dir: root,
+            is_server,
+            requests: RequestMap::new(),
+            streams: StreamMap::new(),
+            uni_stream_ledger: Vec::new(),
```

Comment:
It doesn't seem like this needs to be a vector, can't it just be the highest stream ID allocated? Also, we might want to add an API to the transport layer to do this (e.g. `open_stream_bidi/uni()` that returns the lowest available stream ID).